lib/pbench/agent/tool_group.py

```diff
@@ -0,0 +1,158 @@
+        """
+        _pbench_run = os.environ.get("pbench_run") if pbench_run is None else pbench_run
+        if not _pbench_run:
+            raise BadToolGroup(
+                "Cannot validate tool group, '{group}', 'pbench_run'"
+                " environment variable missing"
+            )
+
+        tg_dir_name = Path(_pbench_run, f"{ToolGroup.TOOL_GROUP_PREFIX}-{group}")
+        try:
+            tg_dir = tg_dir_name.resolve(strict=True)
+        except FileNotFoundError:
+            raise BadToolGroup(
+                f"Bad tool group, '{group}': directory {tg_dir_name} does not exist"
+            )
+        else:
+            if not tg_dir.is_dir():
+                raise BadToolGroup(
+                    f"Bad tool group, '{group}': directory {tg_dir_name} not valid"
+                )
+            else:
+                return tg_dir
+
+    def __init__(self, group):
+        """Construct a ToolGroup object from the on-disk data of the given
+        tool group.
+
+        If the given tool group is valid, the contents are read into the three
+        dictionary structures:
+
+          "toolnames" - each tool name is the key, with separate dictionaries
+          for each registered host
+
+          "hostnames" - each registered host is the key, with separate
+          dictionaries for each tool registered on that host
+
+          "labels" - each registered host name, that has a label, is the key,
+          and the label as the value; if a host is not labeled, it does not
+          show up in this dictionary
+
+        Raises BadToolGroup via the verify_tool_group() method on error.
+        """
+        self.tg_dir = self.verify_tool_group(group)
+        self.group = group
+
+        # __trigger__
+        try:
+            _trigger = (self.tg_dir / "__trigger__").read_text()
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                raise
+            # Ignore missing trigger file
+            self.trigger = None
+        else:
+            if len(_trigger) == 0:
+                # Ignore empty trigger file contents
+                self.trigger = None
+            else:
+                self.trigger = _trigger
+
+        # toolnames - Dict with tool name as the key, dictionary with host
+        # names and parameters for each host
+        self.toolnames = {}
+        # hostnames - Dict with host name as the key, dictionary with tool
+        # names and parameters for each tool
+        self.hostnames = {}
+        self.labels = {}
+        for hdirent in os.listdir(self.tg_dir):
+            if hdirent == "__trigger__":
+                # Ignore handled above
+                continue
+            if not (self.tg_dir / hdirent).is_dir():
+                # Ignore wayward non-directory files
+                continue
+            # We assume this directory is a hostname.
+            host = hdirent
+            if host not in self.hostnames:
+                self.hostnames[host] = {}
```

Comment:
It seems like `host` would never be in `self.hostnames`...what am I missing?

Also, is this attribute ever used?